Support V2 DTOs in VimeoClient

anthonycr opened the pull request:
# Summary
This PR adds support for requesting the new DTOs using the `VimeoClient`. All usage of the V1 DTOs has been removed in the library, except that of `VimeoError`. Future work will change over to using the `VimeoResponse.Error` class, but for now, we will rely on `VimeoError` since the conversion is not entirely straight forward.

In addition to adding support for these new DTOs, a few changes were made to the V2 DTOs, particularly adding `Serializable` to DTOs that were missing it, adding DTOs and properties that were missing, and redesigning the `VideoFile` class so that it supports differentiation between HLS and DASH.

- [x] Change target branch to `develop` after https://github.com/vimeo/vimeo-networking-java/pull/422 is merged.

anthonycr committed:
Support new DTOs only in the networking library

anthonycr committed:
Merge branch 'develop' into new-dto-support

anthonycr committed:
Fixing tests and example

anthonycr committed:
Adding missing privacy type

anthonycr committed:
Adding missing DashVideoFile DTO and renaming VideoFile to HlsVideoFile, and renaming PlayableFile to VideoFile

anthonycr committed:
Adding missing DRM DTO and play property

anthonycr committed:
Making position property nullable

anthonycr committed:
Adding the album interaction

anthonycr committed:
Rename clip property for consistency with other "clip" properties

anthonycr committed:
Adding missing user property used in context of a video

anthonycr committed:
Updating PrivacySetingsParams with new DTOs

anthonycr committed:
Adding missing password property

anthonycr committed:
Adding missing VideoSeasonConnection

anthonycr committed:
Merge branch 'develop' into new-dto-support

anthonycr committed:
Add missing serializable implementation and fixing test

anthonycr committed:
Fix upload class since gcs is actually a list

anthonycr committed:
Adding missing bypassToken property

anthonycr committed:
Add missing URI property

anthonycr committed:
Fixing channel time types

anthonycr committed:
Fixing group privacy property name

anthonycr committed:
Fixing type of secondsRemaining property

anthonycr committed:
Fixing the types of size properties

anthonycr committed:
Exposing moshi builder

anthonycr committed:
Parsing VimeoError using Gson

anthonycr committed:
Correct json property name

anthonycr committed:
Merge branch 'updated-models' into new-dto-support

anthonycr committed:
Merge branch 'updated-models' into new-dto-support

anthonycr committed:
Change out hash map for map to support moshi

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/utils/VimeoNetworkUtil.java:
```diff
@@ -197,10 +223,7 @@ public void run() {
-                        .getRetrofit()
-                        .responseBodyConverter(VimeoError.class, new Annotation[0]);
-                vimeoError = errorConverter.convert(response.errorBody());
+                vimeoError = getGson().fromJson(response.errorBody().string(), VimeoError.class);
```
I am seeing errors upon log out in the app. Could be related?
```
2020-03-30 15:16:38.712 18181-18181/com.vimeo.android.videoapp E/BaseStreamFragment BaseStreamFragment.onError (643): error in getFetchCallback - Retrofit - Authorization Required
    com.vimeo.networking.model.error.VimeoError
        at com.vimeo.networking.model.error.VimeoError$TypeAdapter.read(VimeoError$TypeAdapter.java:80)
        at com.vimeo.networking.model.error.VimeoError$TypeAdapter.read(VimeoError$TypeAdapter.java:13)
        at com.google.gson.Gson.fromJson(Gson.java:927)
        at com.google.gson.Gson.fromJson(Gson.java:892)
        at com.google.gson.Gson.fromJson(Gson.java:841)
        at com.google.gson.Gson.fromJson(Gson.java:813)
        at com.vimeo.networking.utils.VimeoNetworkUtil.getErrorFromResponse(VimeoNetworkUtil.java:226)
        at com.vimeo.networking.callbacks.VimeoCallback.onResponse(VimeoCallback.java:53)
        at retrofit2.DefaultCallAdapterFactory$ExecutorCallbackCall$1.lambda$onResponse$0$DefaultCallAdapterFactory$ExecutorCallbackCall$1(DefaultCallAdapterFactory.java:81)
        at retrofit2.-$$Lambda$DefaultCallAdapterFactory$ExecutorCallbackCall$1$3wC8FyV4pyjrzrYL5U0mlYiviZw.run(Unknown Source:6)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

anthonycr committed:
Merge branch 'updated-models' into new-dto-support

anthonycr committed:
Removing unnecessary annotation

anthonycr committed:
Compare string values instead of enum values

anthonycr committed:
Fixing parameter type for Moshi

WhosNickDoglio commented on vimeo-networking/src/main/java/com/vimeo/networking/AnnotatedConverterFactory.java:
```diff
@@ -90,10 +90,10 @@
     @NotNull
     private Converter.Factory chooseFactoryForSerializer(@NotNull final Annotation[] annotations) {
         final Serializer serializer = findSerializerInArray(annotations);
-        if (serializer != null && serializer.converter() == ConverterType.MOSHI) {
```
Do we still need this class? Don't all the new models use Moshi?

anthonycr replied:
I'm going to remove all the V1 models and associated code in an upcoming ticket.

WhosNickDoglio commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java:
```diff
@@ -247,7 +245,7 @@ public void setVimeoAccount(@Nullable VimeoAccount vimeoAccount) {
             // If the provided account was null but we have an access token, persist the vimeo account with
             // just a token in it. Otherwise we'll want to leave the persisted account as null.
-            vimeoAccount = new VimeoAccount(mConfiguration.mAccessToken);
+            vimeoAccount = new VimeoAccount(mConfiguration.mAccessToken, null, null, null, null, null);
```
`@JvmOverloads` on a data class constructor? Good or bad idea?

Or something like `VimeoAccountFactory.createAccountWithAccessToken(mConfiguration.mAccessToken)` ?

WhosNickDoglio commented on vimeo-networking/src/main/java/com/vimeo/networking/utils/VimeoNetworkUtil.java:
```diff
@@ -86,6 +93,23 @@ public static Gson getGson() {
+     * @return Moshi object that can be used to serialize and deserialize JSON.
+     */
+    @NotNull
+    public static Moshi getMoshi() {
```
Shouldn't `RetrofitSetup` use this method?

WhosNickDoglio commented on models/src/main/java/com/vimeo/networking2/params/ModifyVideosInAlbumSpecs.kt:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * The set of videos that should be added.
+     */
+    @Json(name = "set")
```
Is this the correct JSON key? Just want to confirm since `ModifyVideoInAlbumsSpecs` has a similar property with the key as "add".

anthonycr replied:
I copied the original Java class, @howardrigberg maybe you can weigh in on why this is named `set` instead of `add`?

howardrigberg replied:
It's how API decided to do it. I believe that I asked the question at the time, but I don't remember the rationale. 🤷‍♂

WhosNickDoglio reviewed (changes requested):
Looking good! Just a few suggestions and question. 

Also can we depreciate or just remove completely all the v1 models?

howardrigberg commented on example-kotlin-android/src/main/kotlin/com/vimeo/android/networking/example/kotlin/AccountPreferenceManager.kt:
```diff
@@ -3,9 +3,9 @@ package com.vimeo.android.networking.example.kotlin
-import com.vimeo.networking.model.VimeoAccount
 import com.vimeo.networking.utils.VimeoNetworkUtil
+import com.vimeo.networking2.User
+import com.vimeo.networking2.VimeoAccount
```
Optional: there's a trivial lint error in this file on line 50 if you want to fix it.

howardrigberg commented on example-kotlin-android/src/main/kotlin/com/vimeo/android/networking/example/kotlin/MainActivity.kt:
```diff
@@ -10,10 +10,9 @@ import com.vimeo.networking.VimeoClient
-import com.vimeo.networking.model.User
-import com.vimeo.networking.model.VideoList
 import com.vimeo.networking.model.error.VimeoError
+import com.vimeo.networking2.User
```
Optional. Some lint warnings in this file if you want to fix them.

howardrigberg commented on models/src/main/java/com/vimeo/networking2/DashVideoFile.kt:
```diff
@@ -0,0 +1,52 @@
+ * A video file that represents a DASH stream.
+ */
+@JsonClass(generateAdapter = true)
+data class DashVideoFile(
```
When are we planning to delete the java model equivalents?

anthonycr replied:
Yes, also mentioned in a reply to nick, there will be a separate ticket for the full removal of the V1 models.

anthonycr committed:
Use VimeoNetworkUtil Moshi definition instead of duplicating it

howardrigberg commented on models/src/main/java/com/vimeo/networking2/Drm.kt:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * The video file containing the info about the DRM protected stream.
+     */
+    @Json(name = "widevine")
```
Do you know it is possible for us to have a problem where models get serialized with the old system and then deserialized with the new system upon update in the app?

anthonycr replied:
As far as I know, there are no issues with serializing from the old models to the new, since the JSON representations are unchanged.

howardrigberg commented on models/src/main/java/com/vimeo/networking2/DashVideoFile.kt:
```diff
@@ -0,0 +1,52 @@
+ * A video file that represents a DASH stream.
+ */
+@JsonClass(generateAdapter = true)
+data class DashVideoFile(
```
There is an inconsistency with respect to the `VideoFile` class. This class does not inherit from it even though it shares members while `HlsVideoFile` is derived from it.

anthonycr replied:
One of the design principles of V2 is that we are not using inheritance if possible. The design of these classes was completely changed, as it was not correctly designed. The `quality` property existed on the `HlsVideoFile` and the `DashVideoFile` even though it was unnecessary (the quality was always `HLS` or `DASH`). If you look at the way the files were used, the `VideoFile` class was used as an interface to accept HLS, DASH, or progressive streams. Adding `DASH` and `HLS` as artificial qualities was just a hacky way to differentiate between stream types. The use of inheritance also obfuscated common properties when viewing the class. Since the type system already has a built-in way of determining whether it's an HLS or DASH stream, by simply using `instanceof` in java and `is` in kotlin, I simplified the designs of these classes and cleaned them up. They still share the common `VideoFile` interface, where the properties are defined. 

The only other common property shared between `DashVideoFile` and `HlsVideoFile` is the `live` property, which was originally shared in the `VideoFile` abstract class. However, this property does not exist on the `ProgressiveVideoFile` DTO, since live streams are always DASH or HLS, so I opted to duplicate it across the two files, as I don't think it warrants creating a separate interface.

anthonycr committed:
Use a factory instead of using the default constructor

howardrigberg commented on models/src/main/java/com/vimeo/networking2/params/AddVideoToAlbum.kt:
```diff
@@ -0,0 +1,22 @@
+     */
+    @Json(name = "position")
+    val position: Int?
+)
```
Should we add `Serializable` for the params?

anthonycr replied:
No, the `VimeoService` now serializes parameters using Moshi (See the modifications to `AnnotatedConverterFactory`)

howardrigberg commented on models/src/main/java/com/vimeo/networking2/params/ModifyVideosInAlbumSpecs.kt:
```diff
@@ -0,0 +1,23 @@
+     * The set of videos that should be removed.
+     */
+    @Json(name = "remove")
+    val removeVideoSet: Set<NamedWrapperForRemove>? = null,
```
The original design of hiding the `NamedWrapper` classes was intentional. It was an implementation detail that is necessary to get the data to serialize to json properly, but it should not be exposed to consumers of the networking library. I wanted to avoid an unnecessary layer of complication in the networking api by exposing this class.

anthonycr replied:
I could make them inner classes, but we can't make them private, because Moshi requires the properties be visible to the adapters (they aren't visible if they are private), and we can't make the properties public but the class type private as that results in a kotlin error. Let me know if you have other ideas about how you want to design this class in Kotlin + Moshi. Personally, it doesn't bother me that these classes are public, but I understand the original design you implemented was user friendly. Perhaps a custom constructor would be suitable to improve the UX. Let me know what you think.

howardrigberg commented on models/src/main/java/com/vimeo/networking2/params/NamedWrapperForAdd.kt:
```diff
@@ -0,0 +1,16 @@
+import com.squareup.moshi.JsonClass
+
+/**
+ * Container for [AddVideoToAlbum].
```
Is there a reason why this class can't be private and internalized within `ModifyVideosInAlbumSpecs`? It is an internal detail to accommodate the api and is not needed outside of the class itself.

anthonycr replied:
Cannot be made `private` as the property using it must be publicly exposed for use with Moshi.

howardrigberg commented on models/src/main/java/com/vimeo/networking2/params/NamedWrapperForRemove.kt:
```diff
@@ -0,0 +1,16 @@
+ * Container for [RemoveVideoFromAlbum].
+ */
+@JsonClass(generateAdapter = true)
+data class NamedWrapperForRemove(
```
Is there a reason why this class can't be private and internalized within `ModifyVideosInAlbumSpecs`? It is an internal detail to accommodate the api and is not needed outside of the class itself.

anthonycr replied:
Cannot be made `private` as the property using it must be publicly exposed for use with Moshi.

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java:
```diff
@@ -762,8 +760,8 @@ public void success(VimeoAccount vimeoAccount) {
-                final String userName = vimeoAccount.getUser().mName;
-                final String name = userName != null ? userName : vimeoAccount.getUser().mUri;
+                final String userName = vimeoAccount.getUser().getName();
+                final String name = userName != null ? userName : vimeoAccount.getUser().getName();
```
This should be `getUri()` not `getName()` at the end of this statement, right?

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java:
```diff
@@ -1483,13 +1482,13 @@ public Call modifyVideoInAlbums(@NotNull final Video video,
      * @param callback The VimeoCallback containing PictureResource data
```
change to `PictureCollection`

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java:
```diff
@@ -1483,13 +1482,13 @@ public Call modifyVideoInAlbums(@NotNull final Video video,
      */
     @Nullable
-    public Call<PictureResource> createPictureResource(String uri, VimeoCallback<PictureResource> callback) {
+    public Call<PictureCollection> createPictureResource(String uri, VimeoCallback<PictureCollection> callback) {
```
Just want to double check that this was intentional and without side-effects. `PictureCollection` and `PictureResource` are technically different objects.  `PictureResource` represents a subset of `PictureCollection` and nullability of members probably means this will function properly, but it is a little confusing. Maybe `PictureCollection` from `PictureResource` in the new models and use  `PictureResource` here to keep things the same?  I also noticed that you did not change the object type from `PictureResource` to `PictureCollection` in `activatePictureResource` below.

anthonycr replied:
I realize that one is a subset of the other, but in the API code, they share the same definition, so I intentionally removed `PictureResource` and replaced it with `PictureCollection`. I will change the function names and docs to reference the picture collection.

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoService.java:
```diff
@@ -425,11 +415,10 @@
-                                                          @Url String uri,
-                                                          @QueryMap Map<String, String> options,
-                                                          @Header("Cache-Control") String cacheHeaderValue);
+    Call<FeedList> getFeedListMoshi(@Header("Authorization") String authHeader,
```
Change the name to `getFeedList`?

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoService.java:
```diff
@@ -462,10 +451,10 @@
+                                   @Header("Cache-Control") String cacheHeaderValue);
 
     @GET
     Call<UserList> getUserList(@Header("Authorization") String authHeader,
```
Remove the `getUserListMoshi` and `getVideoListMoshi` methods?

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoService.java:
```diff
@@ -551,13 +540,13 @@
                       @Url String uri,
                       @Header("Cache-Control") String cacheHeaderValue,
-                      @Body HashMap<String, String> parameters);
+                      @Body Map<String, String> parameters);
```
Is this version of `POST` being used? Can we delete it? Ditto for the `GET`.

howardrigberg commented on vimeo-networking/src/main/java/com/vimeo/networking/VimeoService.java:
```diff
@@ -294,15 +291,15 @@
 
     @POST
-    Call<PictureResource> createPictureResource(@Header("Authorization") String authHeader, @Url String uri);
+    Call<PictureCollection> createPictureResource(@Header("Authorization") String authHeader, @Url String uri);
```
See comments elsewhere about `PictureCollection` vs. `PictureResource`.

anthonycr committed:
Use the URI as a backup if the name is null

anthonycr committed:
Updating function names and docs for new DTOs

anthonycr committed:
Remove dead code

anthonycr commented:
@howardrigberg I addressed the issues you pointed out. The only thing outstanding is your comments on the album parameters classes, but I need input from you before I decide to make any changes.

anthonycr committed:
Renaming wrapper classes and adding secondary constructors to avoid redundancy

anthonycr commented:
@howardrigberg changes made, back to you for review.

anthonycr committed:
Fixing bug where occassionally the accessToken is null and it was crashing
